### PR TITLE
Add a hook for creating a new form

### LINKED ISSUE #11: Usage with Vue 3
When using the new Vue 3 composition API, the `setup` function [doesn't have access to `this`](https://v3.vuejs.org/guide/composition-api-setup.html#usage-of-this), which makes it impossible to use the form helpers.

To work around this, a `useForm` hook could be introduced that creates a new `InertiaForm` object when called.

### PR DESCRIPTION
**Pull request by m1guelpf**
Looking into the core of Inertia, we might be fine by using a global Inertia instance.

Closes #11 

Opened the PR in draft mode because as of right now, we're introducing a dependency on `@inertiajs/inertia-vue3` which is not ideal, but I can't think of a way to get access to the `usePage()` hook without doing so. Ideas welcome.

### CODE REVIEW

**Comment by driesvints** [comment]
Is this backwards compatible with VueJS v2? And current Jetstream apps?

**Comment by m1guelpf** [comment]
@driesvints Since I'm just adding a new method, this should be completely backwards-compatible (when I finish removing the dependency on the vue3 Inertia adapter). The only catch is that users with the vue2 adapter will get an error if they try to use the hook.

**Comment by driesvints** [comment]
@m1guelpf I guess that's fine as this is Vue v3 specific.

**Comment by m1guelpf** [comment]
I've tested this on a Jetstream vue3 app and a Jestream vue2 app and everything seems to work properly, although I'd appreciate some more vue2 testing before merging this

**Comment by taylorotwell** [comment]
@claudiodekker this look OK to you?

**Comment by claudiodekker** [comment]
@taylorotwell Yep, looks good to me on first glance 👍 

@m1guelpf I'm assuming you tried `import { usePage } from "@inertiajs/inertia-vue3"` before the current `require ...`? 
(Require loads the whole thing even if you only really need one part, but I'm not sure if my suggestion works inline as well)

File: src/InertiaForm.js [hunk 1]
```diff
@@ -259,6 +260,20 @@ class InertiaForm {
...
+
+	if (!adapter) throw Error('The useForm hook can only be used when @inertiajs/inertia-vue3 is installed.')
+
+    const { props } = adapter.usePage()
+
+    return InertiaForm.create()
+            .withData(data)
+            .withOptions(options)
+            .withInertia(Inertia)
+            .withPage(() => props.value)
```
Comment (claudiodekker) [inline src/InertiaForm.js]: Looking at this a bit closer, this line scares me a bit. 

I honestly haven't had much time yet to look into Vue 3 and our Vue 3 adapter, but unless there's something that I'm missing/not aware of, won't this only load the `value` prop?

**Comment by m1guelpf** [comment]
@claudiodekker Imports only work on the top level, and we want this package to work with people using vue 2 too, so that's why there's a conditional and I've added the package to `optionalDependencies`. Since we need to make a null check, we can't use restructuring as far as I'm concerned.

**Comment by m1guelpf** [comment]
Going back to draft since I've found some issues related to displaying success state

File: src/InertiaForm.js (hunk 1 above)
Comment (m1guelpf) [reply src/InertiaForm.js]: Inertia actually wraps the props object on a `Proxy`, which requires us to call `.value` to get a reference to the object. See https://v3.vuejs.org/guide/composition-api-introduction.html#standalone-computed-properties

**Comment by m1guelpf** [comment]
I've managed to solve the reactivity issue by wrapping the `InertiaForm` object on a `shallowReactive` call (which basically adds reactivity to the object without wrapping it in a proxy (see above), which should make this work and also not crash Inertia. The issue is that this, again, introduces a hard dependency on vue 3. Looking into how to fix that now.

**Comment by m1guelpf** [comment]
Tested on a Vue 2 app and it doesn't seem to cause any issues, so I'd say it's ready to go 🚀

**Review by claudiodekker — APPROVED** [review]

File: src/InertiaForm.js [hunk 2]
```diff
@@ -259,6 +261,20 @@ class InertiaForm {
     }
 }
 
+export function useForm(data, options) {
+    const adapter = require('@inertiajs/inertia-vue3')
+
+	if (!adapter || !shallowReactive) throw Error('The useForm hook can only be used with Vue 3 and  @inertiajs/inertia-vue3.')
```
Comment (driesvints) [inline src/InertiaForm.js]: indentation is off here.

**Pull request merged**